File: src/main/java/com/lmp/shared/pricing/PricingContext.java
package com.lmp.shared.pricing;

import java.math.BigDecimal;

/**
 * Contexte de tarification résolu pour une requête donnée.
 *
 * @param countryCode      Code pays ISO 3166-1 alpha-2 (ex. CA, FR).
 * @param currency         Code devise ISO 4217 (ex. CAD, EUR).
 * @param eurToTargetRate  Taux effectif EUR → {currency} (taux brut × (1 + marge)).
 * @param rawRate          Taux brut depuis Frankfurter (ou statique), sans marge.
 * @param fxMargin         Marge appliquée (0 si aucune).
 * @param rateSource       "frankfurter" ou "static".
 */
public record PricingContext(
        String countryCode,
        String currency,
        BigDecimal eurToTargetRate,
        BigDecimal rawRate,
        BigDecimal fxMargin,
        String rateSource
) {
    /** Rétrocompatibilité : constructeur sans metadata FX. */
    public PricingContext(String countryCode, String currency, BigDecimal eurToTargetRate) {
        this(countryCode, currency, eurToTargetRate, eurToTargetRate, BigDecimal.ZERO, "static");
    }
}

File: backend-dotnet/src/Lmp.Application/Pricing/PricingContext.cs
namespace Lmp.Application.Pricing;

/// <summary>
/// Resolved pricing context for a request. Faithful port of
/// <c>com.lmp.shared.pricing.PricingContext</c>.
/// </summary>
/// <param name="CountryCode">ISO 3166-1 alpha-2 country code (e.g. CA, FR).</param>
/// <param name="Currency">ISO 4217 currency code (e.g. CAD, EUR).</param>
/// <param name="EurToTargetRate">Effective EUR → currency rate (raw rate × (1 + margin)).</param>
/// <param name="RawRate">Raw rate from the FX source, before margin.</param>
/// <param name="FxMargin">Applied FX margin (0 when none).</param>
/// <param name="RateSource">"frankfurter", "fixed" or "static".</param>
public sealed record PricingContext(
    string CountryCode,
    string Currency,
    decimal EurToTargetRate,
    decimal RawRate,
    decimal FxMargin,
    string RateSource)
{
    /// <summary>Backwards-compatible factory without FX metadata.</summary>
    public static PricingContext Of(string countryCode, string currency, decimal eurToTargetRate)
        => new(countryCode, currency, eurToTargetRate, eurToTargetRate, 0m, "static");
}
